File: Files_On_RPi/coop_con_loc_RPi/src/coop_con_loc/src/localization_main.py
#!/usr/bin/env python3
import rospy
import time
import numpy as np
import json
from std_msgs.msg import String
from termcolor import colored
from localization import localization
#from localization_ML import localization
#from ekf import EKF
from ekf_dist import EKF_DIST
#from ekf_dist_jerk import EKF_DIST

Results_File_py = open(r"/root/logs/Results_File_py.txt","w+")
#Results_File_EST = open(r"/root/coop_con_loc_RPi/src/coop_con_loc/src/Results_File_EST.txt","w+")
#Results_File_Freq_Pos = open(r"/root/coop_con_loc_RPi/src/coop_con_loc/src/Results_File_Freq_Pos.txt","w+")
#Results_File_NED_LOC = open(r"/root/coop_con_loc_RPi/src/coop_con_loc/src/Results_File_NED_LOC.txt","w+")

data_Pos_UWB = {}
data_Vel_NED = {}
data_NED = {}
data_ACC = {}
data_DIST = {}

#------------------------------------------------------------------------
#Anchors Information
Num_Anchors = 8
All_Anchors = np.zeros([3,Num_Anchors])
All_Anchors[0:3,0] = [-500,0,250] #A
All_Anchors[0:3,1] = [4880,0,250] #B
All_Anchors[0:3,2] = [0,5920,280] #C
All_Anchors[0:3,3] = [5590,5790,400] #D
All_Anchors[0:3,4] = [0,0,2290] #E
All_Anchors[0:3,5] = [4880,0,2240] #F
All_Anchors[0:3,6] = [0,5920,2390] #G
All_Anchors[0:3,7] = [5260,5790,2290] #H
#-----------------------------------------------------------------------

def clean_shutdown():
	Results_File_py.close()
#	Results_File_EST.close()
#	Results_File_Freq_Pos.close()
#	Results_File_NED_LOC.close()
	print(colored("localization stopped!","green"))

def callback_ACC(data):
	global data_ACC
	data_ACC = json.loads(data.data)

def callback_DIST(data):
	global data_DIST
	data_DIST = json.loads(data.data)

def callback_Vel_NED(data):
	global data_Vel_NED
	data_Vel_NED = json.loads(data.data)

def localization_main():

	DIST = [0]*Num_Anchors
	X = np.zeros([3,1])
	pos_z_Flt = 0.0
	X_m = np.zeros([3,1])
	X_new = np.zeros([3,1])
#	pos_NED_m = np.zeros([3,1])
#	vel_NED_m = np.zeros([3,1])
	Acc_m = np.zeros([3,1])
	time_stmp = 0.0
	x_hat_KF_z = np.zeros([3,1])
	P_KF_z = np.eye(3)
#	x_hat_KF_z = np.zeros([4,1]) # Using KF based on constant jerk
#	P_KF_z = np.eye(4) # Using KF based on constant jerk
	x_hat_EKF = np.zeros([6,1])
	P_EKF = np.eye(6)
	publish_data_NED = String()
	publish_data_Pos_UWB = String()
	psi_coordinate_conversion = 35.0 * np.pi / 180.0 #angle between North and X axis of the UWB system

	Diff_Accept_XY = 2000
	Diff_Accept_Z = 2000

	rospy.init_node("localization_py",anonymous=True)
	rate = rospy.Rate(30)
	rospy.on_shutdown(clean_shutdown)
	pub_Pos_UWB = rospy.Publisher("Pos_UWB", String, queue_size = 20)
	pub_NED = rospy.Publisher("NED_POS_VEL", String, queue_size = 20)
	while not rospy.is_shutdown():
		rate.sleep()
		#--------------------------------------------------------------------------------------------------------------------
		#(START) subscribing to all collected distances
		rospy.Subscriber("ALL_DIST", String, callback_DIST)
		if not data_DIST:
			print("No distances are received!")
		else:
			DIST = data_DIST.get("DIST")
		#(END) subscribing to all collected distances
		#--------------------------------------------------------------------------------------------------------------------
		#(START) measuring the frequency of the localziation estimation process
		Period = float(rospy.get_time()) - time_stmp
		Frequency = 1.0/Period
		print(colored(("Frequency of position estimation: %f Hz" %Frequency),"green"))
#		Results_File_Freq_Pos.write("%.2f\r\n" % Frequency)
		Period_EKF = Period
		if (time_stmp<1.0):
			Period_EKF = 0.0
		time_stmp = float(rospy.get_time())
		#(END) measuring the frequency of the localziation estimation process
		#---------------------------------------------------------------------------------------------------------------------
		#(START) doing localization
		X_new = localization(DIST,X,Num_Anchors,All_Anchors)
		if ((X[0]==0) and (X[1]==0) and (X[2]==0)):
			X = X_new
		if ((abs(X[0]-X_new[0])<Diff_Accept_XY) and (abs(X[1]-X_new[1])<Diff_Accept_XY) and (abs(X[2]-X_new[2])<Diff_Accept_Z)):
			X = X_new
		#(END) doing localization
		#---------------------------------------------------------------------------------------------------------------------
		#(START) Kalman filter on z-axis estimation
		Period_KF = Period_EKF
		x_hat_KF_z,P_KF_z = EKF_DIST(X[2],x_hat_KF_z,P_KF_z,Period_KF)
		pos_z_Flt = x_hat_KF_z[0]
		#(END) Kalman filter on z-axis estimation
		#--------------------------------------------------------------------------------------------------------------------
		#(START) logging localization data
		Results_File_py.write("%d %d %d %d\r\n" % (rospy.get_time(),X[0],X[1],X[2])) #pos_z_Flt
		#print(colored(("Estimated Position(mm): %d %d %d" % (X[0],X[1],pos_z_Flt)),"green"))
		X_m[0] = 0.001*X[0]
		X_m[1] = 0.001*X[1]
		X_m[2] = 0.001*pos_z_Flt #X[2]
		#(END) logging localization data
		#--------------------------------------------------------------------------------------------------------------------
		#(START) publishing estimated local position in UWB frame
#		data_Pos_UWB = {"x_UWB_mm": X[0][0], "y_UWB_mm": X[1][0], "z_UWB_mm": pos_z_Flt[0]} #X[2][0]
#		publish_data_Pos_UWB.data = json.dumps(data_Pos_UWB)
#		pub_Pos_UWB.publish(publish_data_Pos_UWB)
		#(END) publishing estimated local position in UWB frame
		#--------------------------------------------------------------------------------------------------------------------
		#(START) converting the estimated UWB position to the local NED frame
#		x_m_NED = (X_m[0][0]*np.cos(psi_coordinate_conversion)) + (X_m[1][0]*np.sin(psi_coordinate_conversion))
#		y_m_NED = (X_m[0][0]*np.sin(psi_coordinate_conversion)) - (X_m[1][0]*np.cos(psi_coordinate_conversion))
#		z_m_NED = -X_m[2][0]
		#(END) converting the estimated UWB position to the local NED frame
		#--------------------------------------------------------------------------------------------------------------------
		#(START) subscribing to the estimated velocity in local NED frame
#		vx_NED_m = 0.0
#		vy_NED_m = 0.0
#		vz_NED_m = 0.0
#		rospy.Subscriber("Vel_NED", String, callback_Vel_NED)
#		if not data_Vel_NED:
#			print("No estimated NED local velocity is received!")
#		else:
#			vx_NED_m = data_Vel_NED.get("vx_NED_m")
#			vy_NED_m = data_Vel_NED.get("vy_NED_m")
#			vz_NED_m = data_Vel_NED.get("vz_NED_m")
		#(END) subscribing to the estimated velocity in local NED frame
		#--------------------------------------------------------------------------------------------------------------------		
		#(START) subscribing to PX4 acceleration measurement
#		rospy.Subscriber("PX4_ACC", String, callback_ACC)
#		if not data_ACC:
#			print("No acceleration data is received!")
#		else:
#			Acc_m[0] = data_ACC.get("ax")
#			Acc_m[1] = data_ACC.get("ay")
#			Acc_m[2] = data_ACC.get("az")
		#(END) subscribing to PX4 acceleration measurement
		#--------------------------------------------------------------------------------------------------------------------
		#(START) extended Kalman filter (EKF)
#		x_hat_EKF,P_EKF = EKF(X_m,Acc_m,x_hat_EKF,P_EKF,Period_EKF)
#		print(colored(("My EKF estimations: %.3f , %.3f , %.3f , %.3f , %.3f , %.3f" % (x_hat_EKF[0],x_hat_EKF[1],x_hat_EKF[2],x_hat_EKF[3],x_hat_EKF[4],x_hat_EKF[5])),"green"))
#		Results_File_EST.write("%.3f %.3f %.3f %.3f %.3f %.3f\r\n" % (x_hat_EKF[0],x_hat_EKF[1],x_hat_EKF[2],x_hat_EKF[3],x_hat_EKF[4],x_hat_EKF[5]))
		#(END) extended Kalman filter (EKF)
		#--------------------------------------------------------------------------------------------------------------------
		#(START) publishing estimated local position and velocity in local NED frame
		#data_NED = {"px":x_hat_EKF[0][0] , "py":x_hat_EKF[1][0] , "pz":x_hat_EKF[2][0] , "vx":x_hat_EKF[3][0] , "vy":x_hat_EKF[4][0] , "vz":x_hat_EKF[5][0]}		
#		Results_File_NED_LOC.write("%d %.3f %.3f %.3f %.3f %.3f %.3f\r\n" % (rospy.get_time(),x_m_NED,y_m_NED,z_m_NED,vx_NED_m,vy_NED_m,vz_NED_m))
#		data_NED = {"px":x_m_NED , "py":y_m_NED , "pz":z_m_NED , "vx":vx_NED_m , "vy":vy_NED_m , "vz":vz_NED_m}
#		publish_data_NED.data = json.dumps(data_NED)
#		pub_NED.publish(publish_data_NED)
		#(END) publishing estimated local position and velocity in local NED frame
		#--------------------------------------------------------------------------------------------------------------------


if __name__ == '__main__':
	try:
		localization_main()
	except (rospy.ROSInterruptException, rospy.ServiceException, rospy.ROSException):
		pass

File: Files_On_RPi/coop_con_loc_nano/src/coop_con_loc/src/localization_main.cpp
#include "ros/ros.h"
#include "std_msgs/String.h"
#include <stdio.h>
#include <string>
#include <math.h>
#include <iostream>
#include <fstream>
#include "json.hpp"
#include "ekf_dist.h"
#include "localization.h"
#include "localization_main.h"

using json = nlohmann::json;

long double abs_ali(long double inp){
    long double out = inp;
    if (inp < 0.0){
	out = -1.0 * out;
    }
    return out;
}

void callbackDIST(const std_msgs::String::ConstPtr& msg)
{
    auto jsonObject = json::parse(msg->data.c_str());
    data_DIST = jsonObject["DIST"].get<std::vector<int>>();
}

void callbackVel(const std_msgs::String::ConstPtr& msg)
{
    auto jsonObject = json::parse(msg->data.c_str());
    data_vx_NED_m = jsonObject["vx_NED_m"].get<long double>();
    data_vy_NED_m = jsonObject["vy_NED_m"].get<long double>();
    data_vz_NED_m = jsonObject["vz_NED_m"].get<long double>();
}

int main(int argc, char **argv)
{
    std::cout.precision(17);
    long double X[3][1] = {0.0,0.0,0.0};
    long double pos_z_Flt = 0.0;
    long double X_m[3][1] = {0.0,0.0,0.0};
    long double X_new[3][1] = {0.0,0.0,0.0};
    long double x_hat_KF_z[3][1] = {0.0,0.0,0.0};
    long double P_KF_z[3][3] = {
        {1.0,0.0,0.0},
        {0.0,1.0,0.0},
        {0.0,0.0,1.0}
    };
    long double vx_NED_m = 0.0;
    long double vy_NED_m = 0.0;
    long double vz_NED_m = 0.0;
    ros::init(argc,argv,"localization_main");
    ros::NodeHandle n;
    ros::Rate loop_rate(120.0);
    ros::Subscriber sub_DIST = n.subscribe("ALL_DIST",1, callbackDIST);
    ros::Subscriber sub_Vel = n.subscribe("Vel_NED",1, callbackVel);
    ros::Publisher pub_Pos_UWB = n.advertise<std_msgs::String>("Pos_UWB",1);
    ros::Publisher pub_NED = n.advertise<std_msgs::String>("NED_POS_VEL",1);
    ros::Time time_stmp = ros::Time::now();
    std::ofstream Results_File;
    Results_File.precision(17);
    if (Results_File.is_open()){
	Results_File.close();
    };
    Results_File.open("/root/logs/Results_File.txt");
    std::ofstream Results_File_NED_LOC;
    Results_File_NED_LOC.precision(17);
    if (Results_File_NED_LOC.is_open()){
	Results_File_NED_LOC.close();
    };
    Results_File_NED_LOC.open("/root/logs/Results_File_NED_LOC.txt");
    std::ofstream Results_File_Freq_Pos;
    Results_File_Freq_Pos.precision(17);
    if (Results_File_Freq_Pos.is_open()){
	Results_File_Freq_Pos.close();
    };
    Results_File_Freq_Pos.open("/root/logs/Results_File_Freq_Pos.txt");

    while (ros::ok())
    {
	long double DIST[Num_Anchors][1];
	//-----------------------------------------------------------------------------------------------
	//(START) subscribing to all collected distances	
	if (data_DIST.empty())
	{
	    for (int i=0;i<Num_Anchors;i++)
	    {
		DIST[i][0] = 0.0;
	    }
	}
	else
	{
	    for (int i=0;i<Num_Anchors;i++)
	    {
		DIST[i][0] = (long double)(data_DIST[i]);
	    }
	}	
	//(END) subscribing to all collected distances
	//-----------------------------------------------------------------------------------------------
	//(START) measuring the frequency of the localziation estimation process
	long double period = (ros::Time::now() - time_stmp).toSec();
	time_stmp = ros::Time::now();
	long double frequency = 1.0/period;
//	std::cout << "Cpp Localization Frequency (Hz):" << ' ' << frequency<< std::endl;
	Results_File_Freq_Pos << time_stmp << " " << frequency << "\n";
	//(END) measuring the frequency of the localziation estimation process
	//-----------------------------------------------------------------------------------------------
	//(START) doing localization
	X_new[0][0] = 0.0;
	X_new[1][0] = 0.0;
	X_new[2][0] = 0.0;
	localization(*DIST,*X_new);
	if ((X[0][0]==0.0) && (X[1][0]==0.0) && (X[2][0]==0.0)){
	    X[0][0] = X_new[0][0];
	    X[1][0] = X_new[1][0];
	    X[2][0] = X_new[2][0];
	};
	long double C_val_x = abs_ali(X[0][0]-X_new[0][0]);
	long double C_val_y = abs_ali(X[1][0]-X_new[1][0]);
	long double C_val_z = abs_ali(X[2][0]-X_new[2][0]);
	if ((C_val_x<Diff_Accept_XY) && (C_val_y<Diff_Accept_XY) && (C_val_z<Diff_Accept_Z)){
	    X[0][0] = X_new[0][0];
	    X[1][0] = X_new[1][0];
	    if (X_new[2][0]>0.0){
		X[2][0] = X_new[2][0];
	    }
	};
	//(END) doing localization
	//---------------------------------------------------------------------------------------------------------------------
	//(START) Kalman filter on z-axis estimation
	ekf_dist(X[2][0],*x_hat_KF_z,*P_KF_z,period,*x_hat_KF_z,*P_KF_z);
	pos_z_Flt = x_hat_KF_z[0][0];
	//(END) Kalman filter on z-axis estimation
	//--------------------------------------------------------------------------------------------------------------------
	//(START) logging localization data
	Results_File << time_stmp << " " << X[0][0] << " " << X[1][0] << " " << pos_z_Flt << "\n";
	X_m[0][0] = 0.001*X[0][0];
	X_m[1][0] = 0.001*X[1][0];
	X_m[2][0] = 0.001*pos_z_Flt; //X[2][0];
	//(END) logging localization data
	//--------------------------------------------------------------------------------------------------------------------
	//(START) converting the estimated UWB position to the local NED frame
	long double x_m_NED = (X_m[0][0]*cos(psi_coordinate_conversion)) + (X_m[1][0]*sin(psi_coordinate_conversion));
	long double y_m_NED = (X_m[0][0]*sin(psi_coordinate_conversion)) - (X_m[1][0]*cos(psi_coordinate_conversion));
	long double z_m_NED = -X_m[2][0];
	//(END) converting the estimated UWB position to the local NED frame
	//--------------------------------------------------------------------------------------------------------------------
	//(START) subscribing to estimated NED velocity
	if (data_vx_NED_m == -100.0)
	{
//	    ROS_INFO("No NED velocities are received!");
	}else
	{
//	    ROS_INFO("NED velocities are received!");
//	    std::cout <<data_vx_NED_m<< std::endl;
	    vx_NED_m = data_vx_NED_m;
	    vy_NED_m = data_vy_NED_m;
	    vz_NED_m = data_vz_NED_m;
	}
	//(END) subscribing to estimated NED velocity
	//--------------------------------------------------------------------------------------------------------------------
	//(START) publishing estimated local position in UWB frame
	json data_pos_UWB;
	data_pos_UWB["x_UWB_mm"] = X[0][0];
	data_pos_UWB["y_UWB_mm"] = X[1][0];
	data_pos_UWB["z_UWB_mm"] = pos_z_Flt; //X[2][0];
	publish_data_Pos_UWB.data = data_pos_UWB.dump();
	pub_Pos_UWB.publish(publish_data_Pos_UWB);
	//(END) publishing estimated local position in UWB frame
	//--------------------------------------------------------------------------------------------------------------------
	//(START) publishing estimated local position and velocity in local NED frame
	Results_File_NED_LOC << time_stmp << " " << x_m_NED << " " << y_m_NED << " " << z_m_NED << " " << vx_NED_m << " " << vy_NED_m << " " << vz_NED_m << "\n";
	json data_NED;
	data_NED["px"] = x_m_NED;
	data_NED["py"] = y_m_NED;
	data_NED["pz"] = z_m_NED;
	data_NED["vx"] = vx_NED_m;
	data_NED["vy"] = vy_NED_m;
	data_NED["vz"] = vz_NED_m;
	publish_data_NED.data = data_NED.dump();
	pub_NED.publish(publish_data_NED);
	//(END) publishing estimated local position and velocity in local NED frame
	//--------------------------------------------------------------------------------------------------------------------
	ros::spinOnce();
	loop_rate.sleep();
    };
    Results_File.close();
    Results_File_NED_LOC.close();
    return 0;
}

